source/language/classical.rst

```diff
@@ -427,6 +427,279 @@ Terminating the program early
 The statement ``end;`` immediately terminates the program, no matter what scope
 it is called from.
 
+
+The Switch statement
+--------------------
+
+A ``switch`` statement is a form of flow control that provides for a predicated selection of zero, one or more statements to be executed based on a discriminating controlling value. The discriminating controlling value can be either *explicit* - as it is the case for ``case`` statements - or *none of the above* - which is the case for ``default`` statements.
+
+A ``switch`` statement is not a loop. It does not iterate over a sequence of values.
+
+``switch`` statements may appear anywhere in a program where statements are allowed.
+
+An OpenQASM3 ``switch`` statement shall use the following keywords:
+
+- ``switch``
+
+- ``case``
+
+- ``default``
+
+- ``break``
+
+
+An OpenQASM3 ``switch`` statement shall be the following grammar:
+
+- The ``switch`` keyword.
+   
+- A right paren ``(`` literal.
+   
+- A ``controlling expression``.
+   
+- A left paren ``)`` literal.
+   
+- A left brace ``{`` literal.
+   
+- A sequence of one or more ``case`` statements (defined below).
+   
+- Either zero or one ``default`` statement(s) (defined below).
+   
+- A right brace ``}`` literal.
+   
+
+The ``controlling expression`` of a ``switch`` statement shall be of integer type. Implicit conversions to an integer type are not allowed.
```

Comment:
Would it make sense to also allow bit array types in controlling expressions? i.e. so that one could have something like:
```
bit[3] r;
qubit[3] q;
for n in [0:2] {
   r[n] = measure q[n];
}
switch (r) {
   case "000":
   case "001":
      // ...
}

Reply:
It's no longer a `switch` statement. It's a collection of `for` loops inside a sequence of `if / else if / else` conditionals.

A `switch` statement is explicitly _not_ a loop.

The whole point of a `switch` statement is the construction of a LUT. That is, the `case` statement label value **_must_** be a known compile-time constant. If the controlling `case` label statement value is not a known compile-time constant, the compiler cannot construct a LUT. That defeats the entire purpose of the `switch` statement.

Also: `case` label values must be integer types, not strings.

Even if we considered the possibility of allowing `switch` on a `String` type, OpenQASM3 does not define what a `String` type is. That being the case, how can we `switch` on it?

The complexity of switching on strings is `O(N)` times number of `case` statements. The compiler will have to write a `strcmp(3C)` for every single `case` statement, and the complexity of `strcmp(3C) is `O(N)`. That's antithetic to the purpose of a `switch` statement.

- What happens in the case where the `string` case label value that's being switched on does not exist?
- How does the compiler check that all the possible `case` label statement values have been exhausted in the `switch` statement?

Reply:
We discussed at TSC meeting today that we can address my usability concern with the following pattern:
```
bit[3] b;
switch (int(b)) {
   case 0b000:
      // ...
   case 0b001:
      // ...
}
```
i.e. write the literals for the cases using OpenQASM's syntax for binary integer literals.

Therefore, we decided to table the idea of extending the allowed types in the controlling statement to include bit arrays. It is something we can revisit in the future.